Smart bulk fhir

adunsulag opened the pull request:
This is WIP pull request to give insight into the Bulk FHIR implementation for multiple patient FHIR api access.  #4203 4203

**Implemented Bulk FHIR Client Credentials Grant**
    
- Extended the SMART app registration to be able to send a JSON Web Key
    Set (JWKS) URI or embed your jwks directly to be able to use system
    permissions.
    
- Implemented a JWT signer using the RS384 algorithm per Bulk FHIR spec
    incorporating code from the jumbojett/OpenID-Connect-PHP project to
    handle the signing.
    
- Implemented a JsonWebKeySet that will retrieve JWKS from a JWK URI or
    hold the JWKS to be used for signature verification.
    
- Wrote Test class to verify the JWK signature verification as well as
    illustrate how to use class mock's to improve unit testing of our
    services.  This will improve testing time in the future as we can proxy
    network requests without actually having to hit the network.

- Implementation of the FHIR Export Controller to handle Bulk FHIR Export.
- Put in the initial $export routes for the Group/Patient/System exports of data for the BULK FHIR implementation.
- Refactored the Capability statement to now support FHIR operations on each FHIR resource.  Moved the Capability statement to use the FHIR class resources for the REST response.  This fixed a number of issues in the FHIR validator for inferno.
- Pulled out the oauth2_trusted_user api calls into its own controller so we can grant a trusted user for Client Credentials grant.
- Refactored dispatch.php in order to allow a trusted client access token that is not tied to a user to still have api access for system calls.
- Refactored the hard dependancy on RestConfig in the ScopeRepository so I could write a unit test against ScopeRepository to fix some scope bugs that were occurring.

Screenshot for System Scope in App registration
![image](https://user-images.githubusercontent.com/228117/106326151-01122380-624a-11eb-902d-ab78c518ffc2.png)

    
We still have outstanding the following requirements for OAUTH2
    authorization.
 - spec requirement to combine and validate keys against both JWK URI &
    client provided keys at registration.  Currently does one or the other.
 - Prevent replay attacks by checking against previous JWT jti values for the
    current issuer (client id) for the 1 minute expires period.
 - Check optional jku value is the same as the JWK URI provided at registration.
 - 
    
That said the current implementation meets all the authorization testing checks of the ONC inferno tool for the Bulk FHIR requirements.

bradymiller commented on _rest_config.php:
```diff
@@ -287,6 +286,29 @@ public static function authorization_check($section, $value): void
+            throw new \OpenEMR\Common\Acl\AccessDeniedException($scopeType, $resource, "scope not in access token");
+        }
+    }
+
```
this block was previously removed

bradymiller replied:
remove block if not being used

bradymiller commented on _rest_config.php:
```diff
@@ -315,6 +337,25 @@ public static function scope_check($scopeType, $resource = null, $permission = n
+            return false;
+        }
+    }
+
```
this block was previously removed

adunsulag replied:
I had some wierd rebase behavior with a bunch of conflicts where things weren't mapping properly.  I'll remove this and the other blocks you mentioned.  Not sure why that happened.

bradymiller replied:
joys of git :)
would remove block if not being used

bradymiller commented on _rest_routes.inc.php:
```diff
@@ -757,14 +769,30 @@
     "GET /fhir/Observation" => function (HttpRestRequest $request) {
-        RestConfig::authorization_check("patients", "med");
-        $return = (new FhirObservationRestController())->getAll($_GET);
+        if ($request->isPatientRequest()) {
```
Will need to make this endpoint patient request proof. I'll add this to my queue.
TODO for me

bradymiller replied:
this is almost completed. i'll plan to bring in that separate PR after this PR is in codebase

bradymiller commented on _rest_routes.inc.php:
```diff
@@ -757,14 +769,30 @@
     "GET /fhir/Observation/:uuid" => function ($uuid, HttpRestRequest $request) {
-        RestConfig::authorization_check("patients", "med");
-        $return = (new FhirObservationRestController())->getOne($uuid);
+        if ($request->isPatientRequest()) {
```
Will need to make this endpoint patient request proof. I'll add this to my queue.
TODO for me

bradymiller replied:
this is almost completed. i'll plan to bring in that separate PR after this PR is in codebase

bradymiller commented on _rest_routes.inc.php:
```diff
@@ -886,6 +914,31 @@
+        RestConfig::apiLog($return);
+        return $return;
+    },
+    'GET /fhir/Patient/$export' => function (HttpRestRequest $request) {
```
looks like this same endpoint is repeated above (guessing you meant to just have it here in this section)

bradymiller replied:
what is a quick example of what $export variable represents (I tried to search for $export in this PR but my brain couldn't figure it out :) )

adunsulag replied:
So the FHIR specification uses the $ sign to mark anything that is in FHIR terminology an [operation](https://www.hl7.org/fhir/operations.html).  Rather than mapping to a Create/Read/Update/Delete request it's supposed to be something that is more of a Remote Procedure Call (RPC) to tell the server to execute something.  In this case the $export is representing the fact that this is the **export-patient** FHIR RPC call to fire off an asynchronous request to export all of the patient data.

bradymiller commented on src/Common/Auth/OpenIDConnect/Grant/CustomClientCredentialsGrant.php:
```diff
@@ -0,0 +1,325 @@
+        // 9. MAC signature verification or REJECT
+        // MAC signature needs to use ES384 or RS384 signature verification @see https://tools.ietf.org/html/rfc7518
+        //
+        // 10. Reject any other invalid JWT per RFC 7519
```
11. Reject if the wind from the north is greater than 5 mph or if keurig stops working :)

adunsulag replied:
Yep, about right.

bradymiller commented on src/Common/Logging/SystemLogger.php:
```diff
@@ -25,6 +25,9 @@ class SystemLogger implements LoggerInterface
     public function __construct()
     {
+        // default to debug, otherwise it's undefined and it stops logging.
+        $logLevel = Logger::DEBUG;
```
this happens when this class is called prior to globals being brought in. this is why i migrated it away from a singleton (then was stuck with the setting even after globals are brought in. below this a placed a section to manually override via comment/uncomment  $logLevel when need debugging prior to globals.

bradymiller replied:
will need to remove above 2 lines

bradymiller commented on tests/Tests/Unit/Common/Auth/OpenIDConnect/Grant/CustomClientCredentialsGrantTest.php:
```diff
@@ -0,0 +1,281 @@
+<?php
```
![2DV](https://user-images.githubusercontent.com/278968/106369572-7a238080-6307-11eb-8846-92414e248be3.gif)

adunsulag replied:
I thought you'd like to see how the PHPUnit mock's work, if you've never used them before.  It makes it so you can avoid HTTP requests in unit tests so that test cases run much faster.  I didn't see them anywhere in the codebase so I put some in, one because it helped with the test, and 2 to illustrate by example how to use them.

adunsulag commented:
Current status.  Still need to write out to the Document storage for the exports and handle the delete operation.  But getting super close.

![image](https://user-images.githubusercontent.com/228117/106695265-3dc27f80-65a8-11eb-8f6c-489ca4650c26.png)

![image](https://user-images.githubusercontent.com/228117/106695295-4a46d800-65a8-11eb-9b6e-1c5446f7d696.png)

![image](https://user-images.githubusercontent.com/228117/106695344-62b6f280-65a8-11eb-8946-cd1f1701d308.png)

bradymiller commented on _rest_config.php:
```diff
@@ -216,20 +216,23 @@ public static function verifyAccessToken()
 
     public static function isTrustedUser($clientId, $userId)
     {
+        $trustedUserService = new \OpenEMR\Services\TrustedUserService();
```
would place this in the `use` block at top and then use short form

adunsulag replied:
ok, this will go up in my next push.

bradymiller commented on _rest_config.php:
```diff
@@ -366,18 +411,30 @@ public static function skipApiAuth($resource): bool
+                    $logResponse = $body->getContents();
+                    $body->rewind();
+                } else {
+                    $logResponse = '';
```
In this case would log a message that not appropriate to log since a binary (then it was logged that something was sent)

adunsulag replied:
Not sure I understand.  Make $logResponse = 'logging skipped as content is not JSON'; or something like that?

bradymiller replied:
yep, then when audit will know that something went out

bradymiller commented on library/classes/Installer.class.php:
```diff
@@ -1157,6 +1165,10 @@ public function quick_install()
                 return false;
             }
+
+            if (! $this->install_additional_users()) {
```
need to also support this in the manual install in setup.php. will basically place the block to support this after line 1231 in setup.php

adunsulag replied:
When is setup.php called versus the installer in the docker setup?  I want to make sure I can test this.

bradymiller replied:
there is a openemr-cmd to reset the openemr install without doing a autoconfigure.
```
openemr-cmd dr
```
Then can install manually via setup.php
(note there are some particular settings needed when installing via setup.php; Server Host is `mysql`, User Hostname is `%`)
(this is something i can also test)

bradymiller replied:
btw, the command to then do an autoinstall (nice to have admin/pass combo for testing) is:
```
openemr-cmd dri
```
and then if want to do an autoinstall with the standard demo data (this is cool because it then tests the sql upgrade script):
```
openemr-cmd drid
```

bradymiller replied:
I should of mentioned that there are commands to also create/use a snapshot (if don't want to lose openemr data and go back to something after testing above; it backs up and restores the openemr data from drive and database). if needed, can even collect it off the docker ("capsule") to use on docker rebuilds.

adunsulag replied:
I knew about the drid, haven't used the dr or dri commands.  I'll have to look at the snapshot stuff as that will be useful.  Thanks.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -199,3 +199,30 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+  `user_id` varchar(40) NOT NULL,
+  `client_id` varchar(80) NOT NULL,
+  `status` varchar(40) NOT NULL,
+  `start_time` datetime NOT NULL,
```
datetime should default to NULL

adunsulag replied:
This is a required field, why would we default to NULL?

adunsulag replied:
This is a required field, why would we default to NULL?

bradymiller replied:
project is in process of supporting sql standards. one of them is the datetime/timestamp fields need to either be a valid date or null (no blank or zero dates). more details here:
https://github.com/openemr/openemr/issues/3378

bradymiller replied:
btw, meant to say mysql strict mode rather than sql standards above

bradymiller replied:
TODO for me. revisit/review why recommending null for empty date

stephenwaite replied:
https://community.open-emr.org/t/value-0000-00-00-is-not-valid-for-mysql-data-type-date-bug-for-mysql-5-5-49-win32/7978/6?u=stephenwaite

bradymiller replied:
thanks for digging up that thread @stephenwaite .

Did a quick test to see what happens when insert columns with nothing in it (except for the `nothing interesting` field where inserted `hey) on the following
![image](https://user-images.githubusercontent.com/278968/106860630-81d18500-6679-11eb-9f20-824511d806ad.png)

The ability to insert 0000-00-00 is definitely problematic since it then needs special care as we convert to a strict mode compatible database. It means in upgrade script we need to change field to allow null, disable strict mode, convert the 0000-00-00, then enable strict mode (note there are examples where we did this in the 6.0.0 upgrade script since a student during gsoc worked on this a bit.

So, for datetime/date/timestamp, we need to keep them at default NULL (ie. can not use NOT NULL).

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -199,3 +199,30 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+  `resource_include_time` datetime DEFAULT NULL,
+  `output_format` varchar(128) NOT NULL,
+  `request_uri` varchar(128) NOT NULL,
+  `resources` text NOT NULL,
```
make text bare:
```
`resources` text,
```

adunsulag replied:
This is also a required field, not sure why we would make it bare?

bradymiller replied:
old mariadb/mysql versions do not support setting of default for text (it just defaults to NULL). There was a large project in past that involved converting these all over to be bare (ie. just default to NULL), so am just keeping this standard going. Since it is a required field (like the date above) these changes should not impact use (since they will never be empty).

bradymiller replied:
TODO for me. revisit/review why recommending bare text

bradymiller replied:
@adunsulag , see the above screenshot for date where I looked at what happens when don't insert anything in a text bare and text not null field. On initial evaluation, this seems ok. However, the problem here is that when users upgrade to 5.0.0, there is code that will basically convert this to bare text (ie. default to null).
https://github.com/openemr/openemr/blob/master/sql/4_2_2-to-5_0_0_upgrade.sql#L89-L92

back in 2015 or so:
https://community.open-emr.org/t/error-when-creating-tables-with-text-columns-with-default-values/7460

fix in 2016:
https://github.com/openemr/openemr/commit/c13ddcbdb71919e3a8223c00c6a28ae3e4c0221d

Because of this, seems like makes most sense to just stick with the pattern of bare TEXT (ie. some users will likely run upgrade inadvertently from older versions and make it that ways anyways :) ) unless there is a big advantage by using NOT NULL.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -199,3 +199,30 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+  `resources` text NOT NULL,
+  `output` text DEFAULT NULL,
+  `errors` text DEFAULT NULL,
+  `access_token_id` text NOT NULL
```
also, for 3 lines above, make text bare

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -199,3 +199,30 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+  `output` text DEFAULT NULL,
+  `errors` text DEFAULT NULL,
+  `access_token_id` text NOT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COMMENT='fhir export jobs';
```
remove charset

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -199,3 +199,30 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+
+#IfNotTable
+CREATE TABLE `export_job` (
+  `uuid` varchar(40) NOT NULL,
```
TODO for me. revisit this after review the export_job mechanism.

bradymiller replied:
Would make uuid binary and a primary key (check out the `uuid_registry` `uuid` column for an example of this). And will show how to create a naturally ordered uuid below where the uuid is made (the smaller size of binary and the natural ordering make it as fast as possible for a uuid). The goal at some point is to use these uuid's instead of incrementing id's to help OpenEMR reach goal of offline support (ie. not always connected to server).

adunsulag replied:
Made this change and pushed it.

bradymiller commented on src/Common/Auth/OpenIDConnect/JWT/JsonWebKeySet.php:
```diff
@@ -0,0 +1,108 @@
+        $this->logger = new SystemLogger();
+        $this->setHttpClient($httpClient);
+
+        $passphrase = '';
```
just wondering if passphrase is gonna be used here?

adunsulag replied:
Yep. The original Key class takes in a passphrase.  The JWKS never uses the passphrase.

bradymiller commented on src/Common/Auth/OpenIDConnect/Repositories/AccessTokenRepository.php:
```diff
@@ -21,6 +21,25 @@
+     */
+    public function getTokenExpiration($tokenId, $clientId, $userId = null)
+    {
+        if (empty($userId)) {
```
Maybe i am misreading the code, but won't all tokens have a user id now that system is binded to the system user?

adunsulag replied:
Yea, good catch.  This was left over from when I didn't have the api system user.

adunsulag replied:
Fixed this.

bradymiller commented on src/Common/Auth/OpenIDConnect/Repositories/ScopeRepository.php:
```diff
@@ -72,15 +158,30 @@ public function finalizeScopes(
     ): array {
+        $finalizedScopes = [];
+        // we only let scopes that the client initially registered with through instead of whatever they request in
+        // their grant.
```
![Uploading 2DV.gif…]()

bradymiller commented on src/Common/Auth/OpenIDConnect/Repositories/ScopeRepository.php:
```diff
@@ -457,6 +573,7 @@ public function getOidcSupportedScopes(): array
         return $this->oidcScopes();
     }
 
+    // TODO: @bradymiller is this used anywhere anymore?
```
not seeing this function used anywhere in master or in this PR, so can remove it

adunsulag replied:
Removed.

bradymiller commented on src/RestControllers/FHIR/FhirExportRestController.php:
```diff
@@ -0,0 +1,491 @@
+                // for now we are just testing with dummy data until we actually run through each resource.
+                $patient = new FHIRPatient();
+                $fhirId = new FHIRId();
+                $fhirId->setValue(Uuid::uuid4());
```
is this uuid the job id or something else? (just trying to figure out best way to incorporate openemr's uuid scheme here)

adunsulag replied:
Fixed this as it was just being used for testing.

bradymiller commented on src/RestControllers/FHIR/FhirExportRestController.php:
```diff
@@ -0,0 +1,491 @@
+     */
+    private function createJobRequest(ExportJob $job)
+    {
+        // we will generate a UUID here, if we ever want the db to do that we would accomplish that here...
```
would use OpenEMR's custom uuid maker, which will create a hybrid ordered/random uuid. Involves:
```
use OpenEMR\Common\Uuid\UuidRegistry;
```
then:
```
$binaryUuid = (new UuidRegistry(['table_name' => 'export_job']))->createUuid();
```

adunsulag replied:
Switched to this format.

adunsulag commented:
@bradymiller @sjpadgett I've gone ahead and pushed the bulk fhir implementation.  All of the infrastructure is there and I implemented the export operation for the Patient and Encounter resources.  I ran into several problems with the US Core Implementation Guide for the Patient resource which I worked on resolving.  There's a number of problems with how the search implementation and some of the US Core resource profiles are implemented and I started going down that rabbit hole, before I decided it'd be best to wrap up where I'm at with things.

Those will need to be fixed before the Bulk FHIR on Inferno can continue.

Best way to test against this is with the Inferno ONC client, but I found a SMART Bulk FHIR client that does some different tests than what Inferno ONC calls for but people can use that if they want to implement Bulk FHIR beyond what ONC requires. The URL is here: https://github.com/adunsulag/bdt/tree/openemr  What I like about this one is that there is no GUI so we could put it into our github actions validators.

After typing this all up it looks like the test suite has some bugs so I'll get those resolved.  BTW I don't know if you check the git commit logs but I'll post my more lengthy feature list from this last commit batch below.
=============================================================================================
Implemented FHIR Document Binary downloads
- Built it so that documents can now be downloaded via FHIR Made it so the document service checks against the category ACLs to allow more fine grained access control on documents. This made it so FHIR export documents require the super user ACL in order to be retrieved from the system.

- Wrote a BaseDocumentDownloader class that is generic and should handle every mime type.  If special mimetype handling is needed someone can implement the IDocumentDownloader interface to handle a specific mime type.

Added an expiration date to Documents
- Documents cannot be downloaded via the FHIR api if they have been marked as deleted or past their expiration date.

- Made additional minor adjustments to the Document class. Added a helper method to Document so you can retrieve the document data using the class method get_data().  It will retrieve it from CouchDB (if enabled) or from the file system. Data is decrypted if you pass the decrypt flag into the function.

- Added a can_access, has_expired, and get_categories helper methods to Document

Built all of the Export FHIR functionality.
- In order for the export to work you have to have the 'Enable OpenEMR Standard FHIR Export API' flag turned on.  By default it is off for security purposes.

- Any FHIR service that wishes to export FHIR resources just needs to implement the IFhirExportableResourceService and reside in the OpenEMR\Services\FHIR namespace.

- Services can state whether they support the patient, group, or system level exports.  Any resource service that is inside the US Core Patient Compartment (https://www.hl7.org/fhir/compartmentdefinition-patient.html) should implement the patient/group export functionality.  Services for FHIR resources outside patient/group should return false for the patient / group methods.

- The FhirPatientService and FhirEncounterService are good reference examples for how to handle an export.  You can send any FHIR resource that extends the FhirResource base class to the ExportWriter and it will handle the data just fine.

- Right now FHIR exports are handled syncronously as the export is all done at the time of the initial export creation.  We assume the OpenEMR server can process this very fast. However, all the code is setup so that this can be moved asynchronously.  Each export has a shutdown time that tracks the last resource exported.  We would need to create a table to store the in progress data but the code should be fairly straightforward to be able to handle this for long running batch jobs.

- The document exports are all handled in memory right now.  If this becomes too much memory usage the ExportStreamWriter is built so it can take a stream.  I intended to add a streaming capability for the Document class but I ran out of time. The export to be more memory efficient can take a stream (such as one from fopen for local file system) and data can be written out that way.  I'm not too sure on how that would work for CouchDB.

Did a bunch of rework on the FhirPatientService for searching.  There's a lot of rework that needs to happen here as we aren't following much
of the specification for searching.  Added in the search parameter types so we can handle our search capability statement correctly.
Also any FHIR resource that is one of the 13 US Core Implementation Guide resources should implement the IResourceUSCIGProfileService so that
we can include the USCIG profiles in the capability statement.

adunsulag commented:
So I figured that it would be good to have a way to test out the client credentials grant without having to download the inferno client.  Creating the assertion is a big pain so I went ahead and wrote a command line test tool to create client credentials grant.  I built a command line framework for executing commands from the OpenEMR namespace and put inside the src/Common/Command folder.

The actually command runner executes from the bin/command-runner file.  That file loads the CommandRunner class which discovers commands, lists them, and executes them based on the command line input of the caller.  This makes it so we can actually write and execute PHP commands to help with development and/or administer the OpenEMR instance (similar to the openemr-cmd utilities).  Currently the CommandRunner doesn't instantiate the system globals as I didn't need it for this test case but someone could extend it to do so if needed.

The command runner only runs in the cli space so it can't be picked up by the php-cgi/apache/nginx.  To list the commands you can execute just run
```
./bin/command-runner -l
```
Which currently outputs:
```
Commands List
Command Usage: command-runner -c <Command> <CommandArguments>
To get help with a command pass the -h flag after the command name

Command Name - Description
CreateClientCredentialsAssertion - utility class to help test and use the client credentials grant assertion
```

To test out the Client Credentials assertion you first need to grab the public JSON Web Key Set to use for registering your client application:
```
./bin/command-runner -c CreateClientCredentialsAssertion -k
```
You can grab the JWKS and register inside the JWKS textarea of the client at <OpenEMR Site Address>/interface/smart/register-app.php If you want to test out the export operations you need to have at least one of the $export operation permssions (system/*.$export, system/Patient.$export, system/Group.$export).  You also need the system/Document.read permission and system/*.$bulkdata-status permission.  If you don't see these permissions you must turn on in Globals -> Connector -> Enable OpenEMR Standard FHIR Export API 

From there you can grab your Client ID for the next step in generating the assertion using the following format:
./bin/command-runner -c CreateClientCredentialsAssertion -i <ClientID> -a <OpenEMR Site Address>/oauth2/default/token
```
./bin/command-runner -c CreateClientCredentialsAssertion -i wg-flmE8OhgaPavLz5LhUdsb1fdhtb-_0wXmGDMVIcA  -a https://localhost:9300/oauth2/default/token
```

That generates a response output that contains the assertion you can use in PostMan/Insomnia/etc as well as a sample CURL request you can immediately execute if you so choose:
```
Executing command 'CreateClientCredentialsAssertion'
Generated Client Credentials Assertion
eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzM4NCJ9.eyJpc3MiOiJ3Zy1mbG1FOE9oZ2FQYXZMejVMaFVkc2IxZmRodGItXzB3WG1HRE1WSWNBIiwiYXVkIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5MzAwXC9vYXV0aDJcL2RlZmF1bHRcL3Rva2VuIiwianRpIjoiN2M4YWQ0ZDktMWNkMy00NDg3LWI1YjAtNGNiZWVmZjc0ZTA0IiwiaWF0IjoxNjEyNjMwNTI1LCJuYmYiOjE2MTI2MzA1MjUsImV4cCI6MTYxMjYzMDU4NSwic3ViIjoid2ctZmxtRThPaGdhUGF2THo1TGhVZHNiMWZkaHRiLV8wd1htR0RNVkljQSJ9.siFIgdpTeyyfAp6HU5j1cwJUkPChhMmxFO5Cc-jQ4oDDz8piY2S_X89S0pUHMczuJvFHcvLrVch1Pr3UjaFjLyRSzsnEczN9xa8gM8ybFCeKw_wTwK5BNdbcJZNNL7RQ1jiZE5PsnICrgV6tR58eS7oVC1k3aEAA-HMXIPLbtmZoRvzz5XOcjajfayBvs0N6-u5QHey2LRKB-iOp3BoCZ_bsMPmhnx33LbgSd46k7CynLGVA8FmkDvWamXdD4XNMPXIB2ie7s9tiqjZREh9MTC80knjLnRw-iPbkkHYyC74HVSwWrksn89TIRRMpaD52kfIck5A6JHJIrzYYXQ-6D_0kofsrv_rFVHa5V2emUqJFUMwU-WPPm8SwFRfE7mgGZlbS8JNFCMi9EY0EFh7r_W7Y47mzA8iMPbTRQiimMuJryVu9wJYDYtCj5qDmn3KBm7Z3auyXPw1BS0Iy4BF0fqCTXJl6YdW10ds1kWCCNF9MWxo9LXtoatZZpX5UgcQO4hqjI9khNDpHnaOwT6sRRdASoTEYVsei5lCirNeAO9fKnGXcOmJvzVydT02GrNO1sC8j3pRPO-07WmWcuOtImSBWJwOOdEU0QArTtYz4QX4ov7MwmdvmbWzl3s5IK_Jc3ATkzle1sR9XeUtMXhhyphprw9FLMM1AU2_Obw3qC2M


Sample CURL request using assertion: 
--> curl -k -X POST --data-urlencode "client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer" \
  --data-urlencode "client_assertion=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzM4NCJ9.eyJpc3MiOiJ3Zy1mbG1FOE9oZ2FQYXZMejVMaFVkc2IxZmRodGItXzB3WG1HRE1WSWNBIiwiYXVkIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5MzAwXC9vYXV0aDJcL2RlZmF1bHRcL3Rva2VuIiwianRpIjoiN2M4YWQ0ZDktMWNkMy00NDg3LWI1YjAtNGNiZWVmZjc0ZTA0IiwiaWF0IjoxNjEyNjMwNTI1LCJuYmYiOjE2MTI2MzA1MjUsImV4cCI6MTYxMjYzMDU4NSwic3ViIjoid2ctZmxtRThPaGdhUGF2THo1TGhVZHNiMWZkaHRiLV8wd1htR0RNVkljQSJ9.siFIgdpTeyyfAp6HU5j1cwJUkPChhMmxFO5Cc-jQ4oDDz8piY2S_X89S0pUHMczuJvFHcvLrVch1Pr3UjaFjLyRSzsnEczN9xa8gM8ybFCeKw_wTwK5BNdbcJZNNL7RQ1jiZE5PsnICrgV6tR58eS7oVC1k3aEAA-HMXIPLbtmZoRvzz5XOcjajfayBvs0N6-u5QHey2LRKB-iOp3BoCZ_bsMPmhnx33LbgSd46k7CynLGVA8FmkDvWamXdD4XNMPXIB2ie7s9tiqjZREh9MTC80knjLnRw-iPbkkHYyC74HVSwWrksn89TIRRMpaD52kfIck5A6JHJIrzYYXQ-6D_0kofsrv_rFVHa5V2emUqJFUMwU-WPPm8SwFRfE7mgGZlbS8JNFCMi9EY0EFh7r_W7Y47mzA8iMPbTRQiimMuJryVu9wJYDYtCj5qDmn3KBm7Z3auyXPw1BS0Iy4BF0fqCTXJl6YdW10ds1kWCCNF9MWxo9LXtoatZZpX5UgcQO4hqjI9khNDpHnaOwT6sRRdASoTEYVsei5lCirNeAO9fKnGXcOmJvzVydT02GrNO1sC8j3pRPO-07WmWcuOtImSBWJwOOdEU0QArTtYz4QX4ov7MwmdvmbWzl3s5IK_Jc3ATkzle1sR9XeUtMXhhyphprw9FLMM1AU2_Obw3qC2M" \
  --data-urlencode "grant_type=client_credentials" \
  --data-urlencode "scope=system/*.\$export system/*.\$bulkdata-status system/Group.\$export system/Patient.\$export system/Encounter.read system/Document.read" https://localhost:9300/oauth2/default/token
```
Note the assertion is only good for one minute so you need to use it and grab your access token quickly.  Your access token is only good for 3 minutes.  You can always generate a new assertion if your time has expired.

Executing the curl command will give you your access token which you can grab to next test out the export operations
```
{"scope":"system\/*.$export system\/*.$bulkdata-status system\/Group.$export system\/Patient.$export system\/Encounter.read system\/Document.read","token_type":"Bearer","expires_in":300,"access_token":"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiJ3Zy1mbG1FOE9oZ2FQYXZMejVMaFVkc2IxZmRodGItXzB3WG1HRE1WSWNBIiwianRpIjoiZmEwMjc2NmFlOWIyN2IyMzMyM2I2ZmMwNDZmODM3MDYzZjcwMmQxM2ZiNjJiMzQ2NTI5NjNiZDVkMDExMzc5Zjc0OWU4MmYzYTM5MGMyNTIiLCJpYXQiOjE2MTI2MzEwMzMsIm5iZiI6MTYxMjYzMTAzMywiZXhwIjoxNjEyNjMxMzMzLCJzdWIiOiI5MmE5YThmZS1kYWJlLTRmOTctYTM5My04MTMwMDUyMTJmYzYiLCJzY29wZXMiOlsic3lzdGVtXC8qLiRleHBvcnQiLCJzeXN0ZW1cLyouJGJ1bGtkYXRhLXN0YXR1cyIsInN5c3RlbVwvR3JvdXAuJGV4cG9ydCIsInN5c3RlbVwvUGF0aWVudC4kZXhwb3J0Iiwic3lzdGVtXC9FbmNvdW50ZXIucmVhZCIsInN5c3RlbVwvRG9jdW1lbnQucmVhZCIsInNpdGU6ZGVmYXVsdCJdfQ.TKFy3wSz57KpeKhY9Fj88mJNsO60hD_ZrJEhPbvOw31WR0AdHJEbmZo0Q7qaJV9LQPS2MKmGbS8_aJQDU67WuDCGj4FiQ-cuFiVogymYIe2rzLRQwkw3d8Zn3HOMpl1FJWWJieehoej3pCnxyHMK9JUtYcw8PPW1i67DmeYyOr0pwURY9C4CaCETU9Y_GeoWQ33f41uBNYQKMAMy6pcYPtrxrh3INIXuey_vg6H29GqtHsMYOgpHzKv_YW4c9hW4sTpLWePohML0d5g5k2HFMH3K6dghWM4gz5KgSJFiKaPtTmqIOWicq4qniyNE67Ae4_hkoJ9WQyjYff-j_eMxvg"}
```

Export Operations have 4 Steps
1. Kick off the export (OpenEMR right now executes the export immediately, but in the future this can just start the request)
2. Poll the export job until the request is complete.
3. Grab the list of files from the export job status request
3. Download each file in a separate FHIR api request

Note if you get an access_denied response it means your access token has expired and you need to generate a new assertion to get a new access token.

To kick off the request you can execute the following CURL command.  If successful the Response Header will contain a 'Content-Location' header telling you the URL you can poll to get status updates on.
We'll use the following command:
```
curl -i -k -H 'Prefer: respond-async' -H 'Accept: application/fhir+json' \
-H 'Authorization: Bearer <AccessToken>' -X GET <OpenEMR Site Address>/apis/default/fhir/\$export
```
We use the -k to disable ssl verification if you are just testing.  -i is to print the headers as we need the Content-Location header for where we go to get status requests. Not we escape the '$' as it has special meaning in a shell command.

For Example:
```
curl -i -k -H 'Prefer: respond-async' -H 'Accept: application/fhir+json' \
-H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiJ3Zy1mbG1FOE9oZ2FQYXZMejVMaFVkc2IxZmRodGItXzB3WG1HRE1WSWNBIiwianRpIjoiNmRlZDMzMjY2Nzk1ZDI0YzAyOTE2MmVmODMyNDczZDkxYmRmMjE5ZWUzODZhNDhkNDNiZmMxMTYzYzFhYzcwMzg3ODc1NTkxNDdiMTRlNWQiLCJpYXQiOjE2MTI2MzE4NTcsIm5iZiI6MTYxMjYzMTg1NywiZXhwIjoxNjEyNjMyMTU3LCJzdWIiOiI5MmE5YThmZS1kYWJlLTRmOTctYTM5My04MTMwMDUyMTJmYzYiLCJzY29wZXMiOlsic3lzdGVtXC8qLiRleHBvcnQiLCJzeXN0ZW1cLyouJGJ1bGtkYXRhLXN0YXR1cyIsInN5c3RlbVwvR3JvdXAuJGV4cG9ydCIsInN5c3RlbVwvUGF0aWVudC4kZXhwb3J0Iiwic3lzdGVtXC9FbmNvdW50ZXIucmVhZCIsInN5c3RlbVwvRG9jdW1lbnQucmVhZCIsInNpdGU6ZGVmYXVsdCJdfQ.VvnD42SeFmZYiHHeLTx0ZkEnJL1kosuMVb_BZgJxk6H0-n1Yd4UzBNSMB1p0_qX5LOFUWzcz8W5GAPiBmKsMQ52yuiHUmxXvmiCNrR6gF5SXYfwMlIzNxC7EN4mfC5z-lCnLUFBmXZ1gct6smCqFwW0YzQFthWYgZ9RwMdCgkEax-HVvI6DbAgIBP7LHIZhRlkJvbX716d4y3Z5DfURDaGXPaX1NhkpL8x6cH3naRnE4GH4bMQ2-7Ec6M71o1GxPOWwWkQvza_E8suWpVj1alSGrsu12BeQXWGpWQ-Avq_pAtQppoRAO66vU381vIqjgIQgG3yuJU8W3HncXM1_bpA' \
-X GET https://localhost:9300/apis/default/fhir/\$export
```
Example Response:
```
HTTP/1.1 202 Accepted
Date: Sat, 06 Feb 2021 17:18:33 GMT
...
Content-Location: https://localhost:9300/apis/default/fhir/$bulkdata-status?job=92ab033c-b93d-472e-97bf-33c5f4a6cbde
Content-Length: 0
Content-Type: application/json
....
```

You can then grab the Content-Location header and execute a curl request against that
```
curl -k -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiJ3Zy1mbG1FOE9oZ2FQYXZMejVMaFVkc2IxZmRodGItXzB3WG1HRE1WSWNBIiwianRpIjoiNmRlZDMzMjY2Nzk1ZDI0YzAyOTE2MmVmODMyNDczZDkxYmRmMjE5ZWUzODZhNDhkNDNiZmMxMTYzYzFhYzcwMzg3ODc1NTkxNDdiMTRlNWQiLCJpYXQiOjE2MTI2MzE4NTcsIm5iZiI6MTYxMjYzMTg1NywiZXhwIjoxNjEyNjMyMTU3LCJzdWIiOiI5MmE5YThmZS1kYWJlLTRmOTctYTM5My04MTMwMDUyMTJmYzYiLCJzY29wZXMiOlsic3lzdGVtXC8qLiRleHBvcnQiLCJzeXN0ZW1cLyouJGJ1bGtkYXRhLXN0YXR1cyIsInN5c3RlbVwvR3JvdXAuJGV4cG9ydCIsInN5c3RlbVwvUGF0aWVudC4kZXhwb3J0Iiwic3lzdGVtXC9FbmNvdW50ZXIucmVhZCIsInN5c3RlbVwvRG9jdW1lbnQucmVhZCIsInNpdGU6ZGVmYXVsdCJdfQ.VvnD42SeFmZYiHHeLTx0ZkEnJL1kosuMVb_BZgJxk6H0-n1Yd4UzBNSMB1p0_qX5LOFUWzcz8W5GAPiBmKsMQ52yuiHUmxXvmiCNrR6gF5SXYfwMlIzNxC7EN4mfC5z-lCnLUFBmXZ1gct6smCqFwW0YzQFthWYgZ9RwMdCgkEax-HVvI6DbAgIBP7LHIZhRlkJvbX716d4y3Z5DfURDaGXPaX1NhkpL8x6cH3naRnE4GH4bMQ2-7Ec6M71o1GxPOWwWkQvza_E8suWpVj1alSGrsu12BeQXWGpWQ-Avq_pAtQppoRAO66vU381vIqjgIQgG3yuJU8W3HncXM1_bpA' \
-X GET https://10.0.0.9:9300/apis/default/fhir/\$bulkdata-status?job=92ab033c-b93d-472e-97bf-33c5f4a6cbde
```
Again note we escape the $bulkdata-status since we are executing in a shell environment.

If the request is successful you'll get the following back as a response:
```
{
  "transactionTime": {
    "date": "2021-02-06 17:18:33.000000",
    "timezone_type": 3,
    "timezone": "UTC"
  },
  "request": "/apis/default/fhir/$export",
  "requiresAccessToken": true,
  "output": [
    {
      "url": "https://localhost:9300/apis/default/fhir/Document/167/Binary",
      "type": "Patient"
    },
    {
      "url": "https://localhost:9300/apis/default/fhir/Document/168/Binary",
      "type": "Encounter"
    }
  ],
  "error": []
}
```

You can then download the generated document.  Each FHIR resource is encoded in JSON and seperated with newlines in the NDJSON format.

Download Document Example:
```
curl -k -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJhdWQiOiJ3Zy1mbG1FOE9oZ2FQYXZMejVMaFVkc2IxZmRodGItXzB3WG1HRE1WSWNBIiwianRpIjoiMGFjMzI4MDYzN2M4ODMzNTc1N2M0NmI0MDIzMGVhNmFhZTlkMGIwODk5MjM5YTcxYzQ2OTI4MWNjNTFjNWVlMTkyNjQ1YTdlNGI1ZGY5NjgiLCJpYXQiOjE2MTI2MzI2MDcsIm5iZiI6MTYxMjYzMjYwNywiZXhwIjoxNjEyNjMyOTA3LCJzdWIiOiI5MmE5YThmZS1kYWJlLTRmOTctYTM5My04MTMwMDUyMTJmYzYiLCJzY29wZXMiOlsic3lzdGVtXC8qLiRleHBvcnQiLCJzeXN0ZW1cLyouJGJ1bGtkYXRhLXN0YXR1cyIsInN5c3RlbVwvR3JvdXAuJGV4cG9ydCIsInN5c3RlbVwvUGF0aWVudC4kZXhwb3J0Iiwic3lzdGVtXC9FbmNvdW50ZXIucmVhZCIsInN5c3RlbVwvRG9jdW1lbnQucmVhZCIsInNpdGU6ZGVmYXVsdCJdfQ.jWbBEXrIbRqZHigeC6X2p-lWqkAe3JAfdGEjjzjlKxa7VxGddRGTW8yS2Ps9ERpnTwvvXCmPwZAbjj0i5jke7UzPGsm_0FthSf0sa62zN67t3U-DEUx0vpP6qJ3EP_jvpzsitFAQATwtlDz7TXtFdpdXPPnY08pm13ziTYLuHsIFwYWFB3lVPLrdMT9BvGhQOkzjv8Ga5WzLGBeq_4swFBTPiTdL-66twSzKUEJCYykTxCNLfRLRcLsgb55oluw1oSXcVLApihH15JAwl0-p9om-xcCMEbgnFGKkDvT9myeSqZHkwcsjNexMm6m1Z85eJWqdA7HV83lv5WygIChnvA' \
-X GET https://localhost:9300/apis/default/fhir/Document/167/Binary
```

Sample Response (This example has some fields removed from the output to improve readability)
```
{"id":"92a9a83f-525e-4710-8716-f9b66c07828b","resourceType":"Patient","name":[{"use":"official","family":"Belford","given":["Phil"]}]}
{"id":"92a9a83f-5484-4a89-87c6-9f43446d3671","resourceType":"Patient","name":[{"use":"official","family":"Underwood","given":["Susan"]}]}
{"id":"92a9a83f-55c3-40d6-baca-49a5ba7e72e1","resourceType":"Patient","name":[{"use":"official","family":"Moore","given":["Wanda"]}]}
```

bradymiller commented on _rest_routes.inc.php:
```diff
@@ -624,14 +627,30 @@
+            // in Patient context you still have to return a bundle even if it is just one resource.
+            $params = $_GET;
+            $params['_id'] = $request->getPatientUUIDString();
+            $return = (new FhirPatientRestController())->getAll($params);
```
TODO for me:
In my ongoing PR, will add a second param in this getAll with puuid to ensure no way to get other patient info

bradymiller replied:
this is almost completed. i'll plan to bring in that separate PR in after this PR is in codebase

bradymiller commented on apis/dispatch.php:
```diff
@@ -177,19 +178,22 @@
         http_response_code(401);
         exit();
     }
+    // if we have no user id but we have a valid token it's because this is a client credentials system request.
```
i don't think this comment is pertinent anymore

bradymiller commented on bin/command-runner:
```diff
@@ -0,0 +1,26 @@
+/**
+ * command-runner.php Handles the loading and running of OpenEMR commands that leverage the php namespaces.
+ * Preference would be to use something like Symfony/Console but we are attempting to avoid too many dependencies so
+ * we have written this simplified command runner for dev & openemr administration needs
```
![2DV](https://user-images.githubusercontent.com/278968/107136184-d8c39e00-68b5-11eb-8667-61ec537a4e64.gif)

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -171,12 +216,91 @@ function __construct($id = "")
+     */
+    public function process_deleted()
+    {
+        // TODO: @bradymiller is this the correct way in the system to delete documents? Do we remove the underlying file?
```
OpenEMR currently deletes the file, but a flag is way better (for logging purposes is very problematic to delete things). So lets definitely go with your mechanism.

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -695,6 +845,48 @@ function createDocument(
+     * @throws BadMethodCallException Thrown if the method is called when the document has been marked as deleted or expired
+     * @return false|string Returns false if the data failed to decrypt, or a string if the data decrypts or is unencrypted.
+     */
+    function get_data($decrypt = false)
```
The natural use when grabbing a document is that it is decrypted if that is the way it is stored. So, if have an override like this, would default it to the standard use case. maybe something like:
```
function get_data($force_no_decrypt = false)
```

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -695,6 +845,48 @@ function createDocument(
+            $couch_revid = $this->get_couch_revid();
+            $couch = new CouchDB();
+            $resp = $couch->retrieve_doc($couch_docid);
+            $data = base64_decode($resp->data);
```
there needs to be a decision point here for docs from couchdb:
1. If the document is encrypted, then decrypt it (or leave as is if the $force_no_decrypt is true)
2. If the document is not encrypted, then do the base64_decode

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -695,6 +845,48 @@ function createDocument(
+            $resp = $couch->retrieve_doc($couch_docid);
+            $data = base64_decode($resp->data);
+        } else {
+            $filepath  = $this->get_url_filepath();
```
if only it were this easy :)
OpenEMR actually did this in the remote past, but the problem is that things then break if the instance is migrated (ie. all the paths will break). Here's an example taken from codebase (with some modes to make it work here):
```
        //change full path to current webroot.  this is for documents that may have
        //been moved from a different filesystem and the full path in the database
        //is not current.  this is also for documents that may of been moved to
        //different patients. Note that the path_depth is used to see how far down
        //the path to go. For example, originally the path_depth was always 1, which
        //only allowed things like documents/1/<file>, but now can have more structured
        //directories. For example a path_depth of 2 can give documents/encounters/1/<file>
        // etc.
        // NOTE that $from_filename and basename($url) are the same thing
        $from_all = explode("/", $filepath);
        $from_filename = array_pop($from_all);
        $from_pathname_array = array();
        for ($i = 0; $i < $this->get_path_depth(); $i++) {
            $from_pathname_array[] = array_pop($from_all);
        }
        $from_pathname_array = array_reverse($from_pathname_array);
        $from_pathname = implode("/", $from_pathname_array);
        $filepath = $GLOBALS['OE_SITE_DIR'] . '/documents/' . $from_pathname . '/' . $from_filename;
```

adunsulag replied:
Thanks for the code snippet.  I integrated it into the Document class.

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -695,6 +845,48 @@ function createDocument(
+            $data = file_get_contents($filepath);
+        }
+
+        if (!empty($data) && $decrypt && $this->is_encrypted()) {
```
as above, note this decrypt stuff needs to already be done above in couchdb, so may make sense to migrate the decrypt into the above if/else block

adunsulag replied:
yep did this.

bradymiller commented on library/globals.inc.php:
```diff
@@ -2961,6 +2961,13 @@ function gblTimeZones()
             xl('Enable OpenEMR Patient Portal RESTful API.')
         ),
 
+        'rest_fhir_export_api' => array(
```
like we discussed in call, would likely be more clear to have an Enable Client Credential grant with a recommendation of something like Turn on only if you know what you are doing :)

adunsulag replied:
So I dug into this more and I misspoke.  Any of the oauth2 grants can grant System permissions.  They can use password grant and a system admin user to get export privielges, they can use the authorization_code and get the system permissions.  What I think we should do is change this global to be 'Enable FHIR System Scopes (Turn on only if you know what you are doing).

This made me realize that I'm not checking the export scopes on each resource to make sure they have permission to access that resource.  I'll add that check in as we still need all of the System Scopes for the FHIR resources we have.

adunsulag replied:
I updated the global flag.  I also added to the FhirExportRestController a check to make sure the user has access to each resource scope in order to export the given resource.  

Found a couple bugs in AuthorizationController as part of this work that you may want to review my latest changes there.

bradymiller commented on setup.php:
```diff
@@ -1229,6 +1219,17 @@ function cloneClicked() {
+                            }
+                            echo "$ok<br />\n";
+                            flush();
                         }
```
Isn't this missing the `install_additional_users()`?

adunsulag replied:
Yes it is, I've added it in my next push, but hopefully you can test this for me, I can't figure out the credentials to use for the setup.php

adunsulag replied:
This should now be fixed, let me know if it works for you.

bradymiller replied:
TODO for me: test the manual install

bradymiller replied:
tested setup.php and it works

![tenor99](https://user-images.githubusercontent.com/278968/107423733-8f4f9a80-6ad1-11eb-88e4-e1f3d6da2531.gif)

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -248,3 +248,43 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+  `output` text,
+  `errors` text,
+  `access_token_id` text,
+  PRIMARY KEY  (`id`)
```
if doing seeks by the uuid, then would set it to a UNIQUE KEY

adunsulag replied:
This is fixed.

bradymiller commented on sql/database.sql:
```diff
@@ -298,6 +298,8 @@ INSERT INTO `categories` VALUES (26, 'Drawings - Eye', '', 17, 47, 48, 'patients
 INSERT INTO `categories` VALUES (28, 'Patient', '', 27, 52, 53, 'patients|docs');
 INSERT INTO `categories` VALUES (29, 'Reviewed', '', 27, 54, 55, 'patients|docs');
+-- @bradymiller is this the right aco for this?  We really don't want people with patient stuff to have access to this doc
+INSERT INTO `categories` VALUES (30, 'FHIR Export Document', '', 1, 55, 56, 'admin|super');
```
```
INSERT INTO `categories` VALUES (30, 'FHIR Export Document', '', 1, 57, 58, 'admin|super');
```
And the number in the categories entry needs to be bumped from 57 to 59

bradymiller replied:
also, the aco looks good

adunsulag replied:
K should be fixed now.

bradymiller replied:
dang, i almost forgot, we also need to change following line in database.sql :)
```
INSERT INTO `categories_seq` VALUES (29);
```
to:
```
INSERT INTO `categories_seq` VALUES (30);
```

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -248,3 +248,43 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+) ENGINE=InnoDB COMMENT='fhir export jobs';
+#EndIf
+
+#IfNotRow categories name FHIR Export Document
```
ok, we have been doing this a bit wrong in the past.
What we need to do is to get the max `rght` value (which will be in the Categories entry btw) in categories and then use that and should look something like this:
```
SET @max_rght = (SELECT MAX(rght) FROM categories);
INSERT INTO categories(`id`,`name`, `value`, `parent`, `lft`, `rght`, `aco_spec`) select (select MAX(id) from categories) + 1, 'FHIR Export Document', '', 1, @max_rght, @max_rght + 1, 'admin|super' from categories where name = 'Categories';
UPDATE categories SET rght = rght + 2 WHERE name = 'Categories';
UPDATE categories_seq SET id = (select MAX(id) from categories);
```

adunsulag replied:
Integrated your suggestions in.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -248,3 +248,43 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+#EndIf
+
+#IfMissingColumn documents date_expires
+ALTER TABLE `documents` ADD COLUMN `date_expires` DATETIME DEFAULT NULL AFTER `date`;
```
remove the `AFTER` clause
(just a potential for breakage in future if we ever for some reason remove the `date` column; better to just have it go to end of table and not worry about placement)

adunsulag replied:
K, fixed this.

bradymiller commented on src/Common/Auth/OpenIDConnect/Grant/CustomClientCredentialsGrant.php:
```diff
@@ -0,0 +1,369 @@
+        if ($userIdentifier === null) {
+            // we want to grab our system user
+            $systemUser = $this->userService->getSystemUser();
+            $userIdentifier = $systemUser['uuid'] ?? null;
```
Just confirming that somewhere in the workflow for the system user (oe-system) that we have ensure it has an assigned uuid? (via call to `(new UuidRegistry(['table_name' => 'users']))->createMissingUuids();`)
I am guessing it is somewhere, but just making sure

bradymiller replied:
and if $userIdentifier is null, then seems best to die

adunsulag replied:
K, I fixed this.

bradymiller commented on src/Common/Auth/OpenIDConnect/Grant/CustomClientCredentialsGrant.php:
```diff
@@ -0,0 +1,369 @@
+            $systemUser = $this->userService->getSystemUser();
+            $userIdentifier = $systemUser['uuid'] ?? null;
+        }
+        $accessToken = parent::issueAccessToken($accessTokenTTL, $client, $userIdentifier, $scopes); // TODO: Change the autogenerated stub
```
Are we ensuring that only those 4 system scopes (i think there were 4) are allowed in this grant request (ie. not allow other system scopes or patient/user scopes)?

adunsulag replied:
I've pushed a fix out that turns off all system scopes unless someone enables the system scope flag.  People can actually request system scopes using Authorization Code grant, Password grant, or Credentials grant.  The systems scope flag handles all those cases.

bradymiller commented on src/Common/Http/HttpRestParsedRoute.php:
```diff
@@ -31,6 +31,13 @@ class HttpRestParsedRoute
      */
     private $resource;
```
PLACEMAKER FOR ONGOING REVIEW

bradymiller commented on src/RestControllers/FHIR/FhirExportRestController.php:
```diff
@@ -0,0 +1,574 @@
+            ->add(new \DateInterval(self::MAX_DOCUMENT_ACCESS_TIME))
+            ->format("Y-m-d H:i:s");
+
+        // set the foreign key so we can track documents connected to a specific export
```
is the foreign key being set to a patient id or something else? If something else, what are some examples?

adunsulag replied:
It's set to the ExportJob's id.  Look at the DeleteRequest in the FhirExportRestController as I need to grab all of the documents for the given ExportJob and mark them all as deleted.

bradymiller replied:
The problem using the foreign_id is that this is basically the pid (aka patient_id) mapping row in the documents table, and codebase and custom code will use this when grabbing documents for a patient. So, can't use it to also map something else (ie. or a patient may inadvertently get a document they should not if the mappings clash). Rec considering making another sql column in documents for this mapping which can include as a parameter in the createDocument call. And would make it a KEY/INDEX.

adunsulag replied:
I wish foreign_id was named patient_id as that is really bizarre naming convention as I assumed foreign_id could point to any other record type in the system.  I guess we can't change it without updating a ton of code that relies on it.  

I will add a `foreign_reference_id` and a `reference_table` column so we know which table we are pointing at.

adunsulag replied:
This should now be fixed.  If you can look over my changes and let me know if anything should be done there, that'd be great.

bradymiller commented on src/RestControllers/FHIR/FhirExportRestController.php:
```diff
@@ -0,0 +1,574 @@
+        if (!empty($result)) {
+            throw new \RuntimeException("Failed to save document for job. Message: " . $result);
+        }
+        // not sure why we have to double save but the document class is broken in handling foreign ids
```
note that when using this mechanism ($folder for patient id in createDocument class  (which is the correct way to do what you are doing), this mechanism will set the foreign id to 0 since it assumes it is not matched to a patient)

adunsulag replied:
I removed the comment since we handle foreign keys now a different way for documents.

bradymiller commented on src/RestControllers/RestControllerHelper.php:
```diff
@@ -12,8 +12,37 @@
+    /**
+     * The resource endpoint names we want to skip over.
+     */
+    const IGNORE_ENDPOINT_RESOURCES = ['.well-known', 'metadata', 'auth'];
```
is auth an endpoint anymore?

adunsulag replied:
I searched _rest_routes.inc.php and didn't see it, so I've gone ahead and removed it.

bradymiller commented on src/Services/FHIR/Document/BaseDocumentDownloader.php:
```diff
@@ -0,0 +1,85 @@
+        // @see Psr\Http\Message\StreamInterface
+
+        $decrypt = true; // decrypt the data if its been encrypted
+        $documentData = $document->get_data($decrypt);
```
as noted above, would assume decrypt rather than need to request it

adunsulag replied:
yep, fixed.

bradymiller commented on src/Services/FHIR/FhirExportJobService.php:
```diff
@@ -0,0 +1,161 @@
+        // back to the caller that something went wrong.
+
+        $ret = sqlInsert($sql, $params);
+        if (!is_int($ret)) {
```
won't this always be an int (since it will return the id created)?

bradymiller replied:
oh wait, I see this means it failed, forget above post

bradymiller commented on src/Services/FHIR/FhirExportJobService.php:
```diff
@@ -0,0 +1,161 @@
+            , $job->getAccessTokenId(), $job->getStatus(), $job->getRequestURI()];
+        // TODO: @bradymiller is there a different function we can call that will NOT kill execution if the insert fails?
+        // for an API caller we'd rather throw an exception or something else rather than just dying so we can communicate
+        // back to the caller that something went wrong.
```
we could make a sqlInsertThrowError function that would instead throw an error rather than die

adunsulag replied:
I went ahead and made that function.

bradymiller commented on tests/Tests/Services/TrustedUserServiceTest.php:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    /**
+     * We had api authorization issues where null user id's were not returning as trusted user's
```
does this function all need to be reworded to system user is rather than null user id?

adunsulag replied:
Yep, it did need to and I fixed it.

bradymiller commented on src/Common/Auth/OpenIDConnect/Repositories/ScopeRepository.php:
```diff
@@ -298,11 +418,13 @@ public function systemScopes(): array
             "system/Consent.read",
```
would it make sense to comment out all the other system scopes like we do for patient (or just remove them since sounds like never plan to implement them)

bradymiller commented:
review complete :)

adunsulag commented:
@bradymiller Thanks brady for the quick turnaround time.  I believe I've fixed all of the issues you identified.  Please let me know if there are additional changes I should make.

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -695,6 +875,87 @@ function createDocument(
+        if (!empty($data)) {
+            if ($this->is_encrypted() && !$force_no_decrypt) {
+                $data = $this->decrypt_content($data);
+            }
```
will need to change above 3 lines to something like this to get couchdb to work correctly:
```
            if ($this->is_encrypted()) {
                    // couchdb is not base64 encoded when it is encrypted, so need to force below flag to false
                    $base64Decode = false;
                    if (!$force_no_decrypt) {
                        $data = $this->decrypt_content($data);
                    }
            }
```

adunsulag replied:
This should now be fixed.

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -695,6 +875,87 @@ function createDocument(
+        $base64Decode = false;
+
+        if ($storagemethod === self::STORAGE_METHOD_COUCHDB) {
+            $base64Decode = true;
```
actually, a clearer way to do what i posted below may be to instead do the following here:
```
if ($this->is_encrypted()) {
    $base64Decode = false;
} else {
    $base64Decode = true;
}
```

adunsulag replied:
This should now be fixed.

bradymiller commented on library/globals.inc.php:
```diff
@@ -2961,6 +2961,13 @@ function gblTimeZones()
         ),
 
+        'rest_system_scopes_api' => array(
+            xl('Enable OpenEMR FHIR System Scopes (Recommended Off, Turn on only if you know what you are doing)'),
```
![tenor (58)](https://user-images.githubusercontent.com/278968/107324614-3a723c80-6a5d-11eb-8661-d7c3b9d1b2d1.gif)

bradymiller commented on library/classes/Document.class.php:
```diff
@@ -207,6 +330,17 @@ function documents_factory($foreign_id = "")
+     * @param string $foreign_id
+     * @return Document[]
+     */
+    public static function getDocumentsForForeignId($foreign_id = "")
```
TODO for me: look into non-patient foreign id use

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -248,3 +248,45 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+    );
+#EndIf
+
+#IfNotTable
```
```
#IfNotTable export_job
```

adunsulag replied:
K this should be fixed now.

bradymiller commented on sql/database.sql:
```diff
@@ -12331,3 +12334,21 @@ CREATE TABLE `x12_remote_tracker` (
+  `output` text,
+  `errors` text,
+  `access_token_id` text,
+  PRIMARY KEY  (`id`)
```
UNIQUE (`uuid`)

(the joys of duplicate tables :) )

adunsulag replied:
K, this should now be fixed.

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -248,3 +248,45 @@ ALTER TABLE `procedure_providers` ADD `type` VARCHAR(31) DEFAULT NULL;
+
+#IfNotRow categories name FHIR Export Document
+SET @max_rght = (SELECT MAX(rght) FROM categories);
+INSERT INTO categories(`id`,`name`, `value`, `parent`, `lft`, `rght`, `aco_spec`) select (select MAX(id) from categories) + 1, 'FHIR Export Document', '', 1, @max_rght, @max_rght + 1, 'admin|super' from categories where name = 'Categories';
```
i think we can now simplify this line to:
```
INSERT INTO categories (`id`,`name`, `value`, `parent`, `lft`, `rght`, `aco_spec`) VALUES ((select MAX(id) from categories) + 1, 'FHIR Export Document', '', 1, @max_rght, @max_rght + 1, 'admin|super');
```

adunsulag replied:
Nope, you get this error: 
```
INSERT INTO categories (`id`,`name`, `value`, `parent`, `lft`, `rght`, `aco_spec`) VALUES ((select MAX(id) from categories) + 1, 'FHIR Export Document', '', 1, @max_rght, @max_rght + 1, 'admin|super')<br />
SQL Error with statement:query failed: INSERT INTO categories (`id`,`name`, `value`, `parent`, `lft`, `rght`, `aco_spec`) VALUES ((select MAX(id) from categories) + 1, &#039;FHIR Export Document&#039;, &#039;&#039;, 1, @max_rght, @max_rght + 1, &#039;admin|super&#039;)--Table &#039;categories&#039; is specified twice, both as a target for &#039;INSERT&#039; and as a separate source for data==&gt;/var/www/localhost/htdocs/openemr/library/sql_upgrade_fx.php at 1175:sqlStatement
```

So I'll leave it be.

bradymiller replied:
oops, my bad on that.

bradymiller commented on src/RestControllers/AuthorizationController.php:
```diff
@@ -18,6 +18,7 @@
 
 use DateInterval;
```
PLACEMARKER FOR ONGOING FINAL REVIEW

bradymiller commented:
@adunsulag , final review complete :)  just a couple minor issues to address and then this is ready to bring in! -brady

adunsulag commented:
@bradymiller All of your review suggestions should now be fixed.  Let me know if you see any more problems.

bradymiller commented:
hi @adunsulag , Code looks good and bringing in. Thanks for this awesome piece of work! -brady

![tenor99](https://user-images.githubusercontent.com/278968/107555922-9b4c6280-6b8c-11eb-8cc7-81b293092543.gif)

![2DV](https://user-images.githubusercontent.com/278968/107555945-a4d5ca80-6b8c-11eb-9a62-8f14c5bfd62c.gif)

sjpadgett commented:
@adunsulag Thanks for the hard work. It's a big deal!

adunsulag commented:
@bradymiller @sjpadgett Thank you both for all of your hard work in helping me get this out.  I recognize there's a lot of room for feature improvement, but I'm very happy for what we were able to accomplish together.
